ulib/FStar.Calc.fst

```diff
@@ -0,0 +1,41 @@
+module FStar.Calc
+
+open FStar.Preorder (* for `relation` *)
+
+noeq
+type calc_proof #t : list (relation t) -> t -> t -> Type =
+  | CalcRefl : #x:t -> calc_proof [] x x
+  | CalcStep : rs:(list (relation t)) -> #p:(relation t) ->
+               #x:t -> #y:t -> #z:t -> calc_proof rs x y -> squash (p y z) -> calc_proof (p::rs) x z
+
+let rec calc_chain_related (#t : Type) (rs : list (relation t)) (x y : t) : Type0 =
+  match rs with
+  | [] -> x == y
+  | r1::rs -> exists w. calc_chain_related rs x w /\ r1 w y
+
+(* Every chain of `t`'s related by `rs` **(reversed!)** has its endpoints related by p *)
+let calc_chain_compatible (#t : Type) (rs : list (relation t)) (p : relation t) : Type0 =
+  forall x y. calc_chain_related rs x y ==> p x y
```

Comment:
We could add a pattern here `calc_chain_related rs x y` (Z3 could also be picking the same automatically).